src/vm/analysis/arithmetic_checker/mod.rs

```diff
@@ -0,0 +1,276 @@
+// Copyright (C) 2013-2020 Blocstack PBC, a public benefit corporation
+// Copyright (C) 2020 Stacks Open Internet Foundation
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+use vm::analysis::types::{AnalysisPass, ContractAnalysis};
+use vm::functions::define::{DefineFunctions, DefineFunctionsParsed};
+use vm::functions::tuples;
+use vm::functions::NativeFunctions;
+use vm::representations::SymbolicExpressionType::{
+    Atom, AtomValue, Field, List, LiteralValue, TraitReference,
+};
+use vm::representations::{ClarityName, SymbolicExpression, SymbolicExpressionType};
+use vm::types::{parse_name_type_pairs, PrincipalData, TupleTypeSignature, TypeSignature, Value};
+
+use std::collections::HashMap;
+use vm::variables::NativeVariables;
+
+pub use super::errors::{
+    check_argument_count, check_arguments_at_least, CheckError, CheckErrors, CheckResult,
+};
+use super::AnalysisDatabase;
+
+#[cfg(test)]
+mod tests;
+
+///
+/// A static-analysis pass that checks whether or not
+///  a proposed cost-function defining contract is allowable.
+/// Cost-function defining contracts may not use contract-call,
+///  any database operations, or iterating operations (e.g., list
+///  operations)
```

Comment:
I think you forgot to mention in this comment that it cannot have any trait definitions or trait references, going by the error variant below.

Reply:
Added in https://github.com/blockstack/stacks-blockchain/pull/2143/commits/c03277838d1e045d9f6c72c2e13e861b90819369